### PR TITLE
Update to latest rustc

### PR DESCRIPTION
**Pull request by DJMcNab**
This had a few complications so there are a handful of workarounds:
 - For some reason, the `get` macro from `rocket` was panicking (at called unwrap on `Option::None` with no further details)
 - However, manually expanding out the `macro`s fixed this, so I did that and removed them
 - The `Frame::loc` field has been made private, although it looks like an accessor will be added
 - rust-analyzer thought `extern crate miri` meant to use the one which was in `rustcSource`, even though that is not correct.

Whilst I'm here, I'm also going to handle the problems that we have from not following `rust_2018_idioms`

Thanks to @bjorn3 for answering my queries :)

This is a very cool project, and I hope it can become more useful. 'Perfect' debugging for rust programs is very compelling, with easy back in time debugging.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -11,15 +11,14 @@ edition = "2018"
...
+rocket = "0.4.7"
+# This is the version of miri for the nightly in the rust-toolchain file. This might be superfluous
+miri = { git = "https://github.com/rust-lang/miri.git", rev = "ea86335318fd06ec964d9a86b187995bda1b6c7d" }
 
 log = "0.4"
 env_logger = "0.7"
 
-serde = "1.0"
-serde_derive = "1.0"
+serde = {version="1.0", features=["derive"]}
```
Comment (bjorn3) [inline Cargo.toml]:
```suggestion
serde = { version="1.0", features=["derive"] }
```

File: src/main.rs [hunk 2]
```diff
@@ -137,26 +118,31 @@ fn create_ecx<'mir, 'tcx>(tcx: TyCtxt<'tcx>) -> InterpCx<'tcx> {
             tracked_call_id: None,
             validate: true,
             stacked_borrows: false,
-            check_alignment: false,
+            // Guesses
```
Comment (bjorn3) [inline src/main.rs]:
```suggestion
```
I think these are good.

File: src/main.rs [hunk 3]
```diff
@@ -137,26 +118,31 @@ fn create_ecx<'mir, 'tcx>(tcx: TyCtxt<'tcx>) -> InterpCx<'tcx> {
...
+            check_alignment: miri::AlignmentCheck::None,
+            track_raw: false,
+            data_race_detector: false,
+            // This is the default, might be better off setting it to 0.0
```
Comment (bjorn3) [inline src/main.rs]: Keeping it at the default should be fine.

File: src/main.rs [hunk 4]
```diff
@@ -137,26 +118,31 @@ fn create_ecx<'mir, 'tcx>(tcx: TyCtxt<'tcx>) -> InterpCx<'tcx> {
             tracked_call_id: None,
             validate: true,
             stacked_borrows: false,
-            check_alignment: false,
+            // Guesses
+            check_alignment: miri::AlignmentCheck::None,
+            track_raw: false,
+            data_race_detector: false,
```
Comment (bjorn3) [inline src/main.rs:123]: Enabling this would likely interfere with the local pretty printing.

File: src/main.rs [hunk 5]
```diff
@@ -403,29 +363,29 @@ fn main() {
...
+
+//     let mut builder = env_logger::Builder::new();
+//     builder.format(format).filter(None, log::LevelFilter::Info);
+
+//     if std::env::var("MIRI_LOG").is_ok() {
+//         builder.parse_filters(&std::env::var("MIRI_LOG").unwrap());
+//     }
+
+//     builder.init();
+// }
```
Comment (bjorn3) [inline src/main.rs]: Please completely remove this.

File: src/render/mod.rs [hunk 6]
```diff
@@ -260,13 +262,25 @@ pub mod routes {
...
-        render::render_main_window(pcx, None, flash.0)
-    });
+    #[get("/")]
+    pub fn index(
+        sender: rocket::State<'_, crate::PrirodaSender>,
+        flash: FlashString,
+    ) -> crate::RResult<Html<String>> {
+        sender.do_work(move |pcx| {
+            let pcx = &*pcx;
+            (|| render::render_main_window(pcx, None, flash.0))()
```
Comment (bjorn3) [inline src/render/mod.rs]: Is this immediately invoked closure necessary? And is the `&*pcx` necessary?

File: src/step.rs [hunk 7]
```diff
@@ -244,124 +244,229 @@ fn parse_def_id(s: &str) -> Result<DefId, String> {
...
-            "already at the start".to_string()
-        }
-    });
+    #[get("/restart")]
+    fn restart(
+        sender: rocket::State<'_, crate::PrirodaSender>,
+    ) -> crate::RResult<rocket::response::Flash<rocket::response::Redirect>> {
+        sender.do_work(move |pcx| {
+            rocket::response::Flash::success(rocket::response::Redirect::to("/"), {
+                pcx.restart();
```
Comment (bjorn3) [inline src/step.rs]: Can you move this out of the `Flash::success` call?

File: src/step.rs [hunk 8]
```diff
@@ -244,124 +244,229 @@ fn parse_def_id(s: &str) -> Result<DefId, String> {
...
         })
-    });
+    }
+
+    #[get("/next")]
+    fn next(
+        sender: rocket::State<'_, crate::PrirodaSender>,
+    ) -> crate::RResult<rocket::response::Flash<rocket::response::Redirect>> {
+        sender.do_work(move |pcx| {
+            rocket::response::Flash::success(rocket::response::Redirect::to("/"), {
```
Comment (bjorn3) [inline src/step.rs]: Same for the block here.

File: src/step.rs [hunk 9]
```diff
@@ -244,124 +244,229 @@ fn parse_def_id(s: &str) -> Result<DefId, String> {
...
-            }
+    }
+    #[get("/single_back")]
+    fn single_back(
+        sender: rocket::State<'_, crate::PrirodaSender>,
+    ) -> crate::RResult<rocket::response::Flash<rocket::response::Redirect>> {
+        sender.do_work(move |pcx| {
+            rocket::response::Flash::success(
+                rocket::response::Redirect::to("/"),
+                (|| {
```
Comment (bjorn3) [inline src/step.rs]: And preferably here if it isn't too hard.

File: src/step.rs [hunk 10]
```diff
@@ -244,124 +244,229 @@ fn parse_def_id(s: &str) -> Result<DefId, String> {
...
+                    )
                 } else {
-                    format!("No breakpoint for for {:?}@{}:{}", breakpoint.0, breakpoint.1.index(), breakpoint.2)
+                    format!("Can't set breakpoint for unwinding without cleanup yet")
                 }
-            }
-            Err(e) => e,
-        }
-    });
+            })
```
Comment (bjorn3) [inline src/step.rs]: And here and below.

File: src/watch/mod.rs [hunk 11]
```diff
@@ -125,55 +126,66 @@ pub fn routes() -> Vec<::rocket::Route> {
...
+pub fn add(
+    sender: rocket::State<'_, crate::PrirodaSender>,
+    id: u64,
+) -> crate::RResult<rocket::response::Flash<rocket::response::Redirect>> {
+    sender.do_work(move |pcx| {
+        rocket::response::Flash::success(rocket::response::Redirect::to("/"), {
+            pcx.traces
+                .alloc_traces
+                .insert(AllocId(id), AllocTrace::new());
+            step_callback(pcx);
```
Comment (bjorn3) [inline src/watch/mod.rs]: And here.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 12]
```diff
@@ -11,15 +11,14 @@ edition = "2018"
 [dependencies]
 regex = "1.3"
 lazy_static = "1.4.0"
-rocket = "0.4.5"
-rocket_codegen = "0.4.5"
-miri = { git = "https://github.com/rust-lang/miri.git", rev = "c641fbde0223cde9fae67de3cc4892998a21cec0" }
+rocket = "0.4.7"
+# This is the version of miri for the nightly in the rust-toolchain file. This might be superfluous
```
Comment (bjorn3) [inline Cargo.toml]:
```suggestion

```
It has to match.

Comment (DJMcNab) [reply Cargo.toml]: We should probably have a comment to that effect then?

Comment (bjorn3) [reply Cargo.toml]:
```suggestion
# This is the version of miri for the nightly in the rust-toolchain file.
```
?

Comment (DJMcNab) [reply Cargo.toml]: I've updated the comment text, although if you'd still prefer to just remove it we can do that too

**Review by bjorn3 — APPROVED** [review]

**Comment by DJMcNab** [comment]
Hopefully a nightly has landed which this PR is compatible with 🎉

I'll check when I'm at a desktop computer.

**Comment by bjorn3** [comment]
My PR missed the deadline for the latest nightly by one merge.

**Comment by DJMcNab** [comment]
It should work now - we've been bitten by `rustc --version` often being a day behind the rustup toolchain. That is, the code runs fine on `nightly-2020-03-01`

**Comment by bjorn3** [comment]
Thanks for the effort!

**Pull request merged**